odbc/src/api/odbc_uri.rs

```diff
@@ -19,23 +38,75 @@ impl<'a> ODBCUri<'a> {
         if odbc_uri.is_empty() {
             return Err(ODBCError::InvalidUriFormat(NOT_EMPTY_ERROR.to_string()));
         }
-        // TODO SQL-990: Support the actual ODBC spec with regards to special characters in attributes
-        // the algorithm will most likely need to be a state machine over each character in the
-        // odbc_uri string.
-        odbc_uri
-            .split(';')
-            .filter(|attr| !attr.is_empty())
-            .map(|attr| {
-                // now split each attribute pair on '='
-                let mut sp = attr.split('=').collect::<Vec<_>>();
-                if sp.len() != 2 {
-                    return Err(ODBCError::InvalidUriFormat(EQUAL_ERROR.to_string()));
-                }
-                // ODBC attribute keys are case insensitive, so we lowercase the keys
-                Ok((sp.remove(0).to_lowercase(), sp.remove(0)))
-            })
-            .collect::<Result<HashMap<_, _>>>()
-            .map(ODBCUri)
+        let mut input = odbc_uri;
+        let mut ret = ODBCUri(HashMap::new());
+        while let Some((keyword, value, rest)) = ODBCUri::get_next_attribute(input)? {
+            ret.0.insert(keyword, value);
```

Comment:
The spec outlines that if keywords are repeated, then the value for the first occurrence of the keyword will be used.  We may want to add a check to only insert if the keywords doesn't already exist here.